tests/unit/stall_detector_test_utilities.hh

```diff
@@ -0,0 +1,124 @@
+            : _old_threshold(engine().get_blocked_reactor_notify_ms())
+            , _old_report(engine().get_stall_detector_report_function()) {
+        engine().update_blocked_reactor_notify_ms(std::chrono::duration_cast<std::chrono::milliseconds>(threshold));
+        engine().set_stall_detector_report_function(std::move(report));
+    }
+
+    ~temporary_stall_detector_settings() {
+        engine().update_blocked_reactor_notify_ms(_old_threshold);
+        engine().set_stall_detector_report_function(std::move(_old_report));
+    }
+};
+
+using void_fn = std::function<void()>;
+
+void spin(std::chrono::duration<double> how_much, void_fn body = []{}) {
+    auto end = internal::cpu_stall_detector::clock_type::now() + how_much;
+    while (internal::cpu_stall_detector::clock_type::now() < end) {
+        body(); // spin!
+    }
+}
+
+[[maybe_unused]]
+static void spin_user_hires(std::chrono::duration<double> how_much) {
+    auto end = std::chrono::high_resolution_clock::now() + how_much;
+    while (std::chrono::high_resolution_clock::now() < end) {
+
+    }
+}
+
+void spin_some_cooperatively(std::chrono::duration<double> how_much, void_fn body = []{}) {
+    auto end = std::chrono::steady_clock::now() + how_much;
+    while (std::chrono::steady_clock::now() < end) {
+        spin(200us, body);
+        if (need_preempt()) {
+            thread::yield();
+        }
+    }
+}
+
+// Triggers stalls by spinning with a specify "body" function
+// which takes most of the spin time.
+[[maybe_unused]]
+static void test_spin_with_body(const char* what, void_fn body) {
+    // The !count_stacks mode outputs stall notification to stderr as usual
+    // and do not assert anything, but are intended for diagnosing
+    // stall problems by inspecting the output. We expect the userspace
+    // spin test to show no kernel callstack, and the kernel test to
+    // show kernel backtraces in the mmap or munmap path, but this is
+    // not exact since neither test spends 100% of its time in the
+    // selected mode (of course, kernel stacks only appear if the
+    // perf-based stall detected could be enabled).
+    //
+    // Then the count_stacks mode tests that the right number of stacks
+    // were output.
+    for (auto count_stacks : {false, true}) {
+        testlog.info("Starting spin test: {}", what);
+        std::atomic<unsigned> reports{};
+        std::function<void()> reporter = count_stacks ? std::function<void()>{[&]{ ++reports; }} : nullptr;
+        temporary_stall_detector_settings tsds(10ms, std::move(reporter));
+        constexpr unsigned nr = 5;
+        for (unsigned i = 0; i < nr; ++i) {
+            spin_some_cooperatively(100ms, body);
+            spin(20ms, body);
+        }
+        testlog.info("Ending spin test: {}", what);
+        BOOST_CHECK_EQUAL(reports, count_stacks ? 5 : 0);
+    }
+}
+
+[[maybe_unused]]
+static void mmap_populate(size_t len) {
```

Comment:
FWIW `static inline` or anonymous namespace removes the need for these `[[maybe_unused]]` annotations.

Reply:
Interestingly enough wrapping the functions in a anonymous namespace will still generate the unused warning. `static line` does work though.

Reply:
Using an anonymous namespace and marking the function as `inline` without static does work though... I guess for both cases since it's a common pattern to have unused `static inline` functions in C that the compiler doesn't give a warning?

Reply:
Ended up using the anon-namespace and marking unused functions as `inline`.

Reply:
This annoyed me enough that I asked a question about it long ago:

https://stackoverflow.com/questions/47542751/gcc-warns-about-unused-static-functions-but-not-static-inline-is-there-a-pract

The basic idea seems to be that this warning seems geared towards static functions in .cc files, not in .h files: having a static function in a .cc file is pointless since it can't be used from the outside. However, it's pretty normal to have functions in header files which aren't used (probably 99.9% of all functions an impl unit header chain aren't going to be used in any given unit) .... but maybe it's bad practice to put static functions in header files.

So then "inline" is just a hint that this probably appears in a header. The discussion on SO has more details.

I think the summary is "use inline" or "use inline in an anonymous namespace" if you really need to share implementations via a header.